winter_sqlalchemy/repository.py

```diff
@@ -0,0 +1,149 @@
+from typing import Iterable
+from typing import Optional
+from typing import TypeVar
+
+from injector import inject
+from sqlalchemy import and_
+from sqlalchemy import exists
+from sqlalchemy import func
+from sqlalchemy import inspect
+from sqlalchemy import select
+from sqlalchemy.engine import Engine
+from sqlalchemy.orm import class_mapper
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.orm.exc import UnmappedClassError
+
+from winter.data import CRUDRepository
+from winter.data.exceptions import NotFoundException
+
+T = TypeVar('T')
+K = TypeVar('K')
+
+
+def sqla_crud(repository_cls):
+    if not issubclass(repository_cls, CRUDRepository):
+        raise TypeError('Repository must be inherited from CRUDRepository before annotating with sqla_crud')
+
+    entity_cls = repository_cls.__entity_cls__
+
+    try:
+        mapper = class_mapper(entity_cls)
+    except UnmappedClassError:
+        raise TypeError('Invalid SQLAlchemy entity class given')
+
+    if len(mapper.tables) > 1:
+        raise TypeError('sqla_crud does not support entities mapped to multiple tables')
+
+    entity_table = mapper.tables[0]
+
+    class RepositoryImpl(repository_cls):
+        """
+        SQLAlchemy implementation for CRUDRepository
+        This repository implementation is not thread-safe.
+        """
+        class RepositoryException(Exception):
+            pass
+
+        def count(self) -> int:
+            return self._engine.execute(select([func.count()]).select_from(entity_table)).scalar()
+
+        def delete(self, entity: T):
+            try:
+                session = self._sessions[entity]
+            except KeyError:
+                raise self.RepositoryException('Entity must be fetched with repository before being deleted')
+            pk = inspect(entity).identity
+            del self._identity_map[pk]
+            session.delete(entity)
+            session.commit()
+            session.close()
+            del self._sessions[entity]
+
+        def delete_many(self, entities: Iterable[T]):
+            for entity in entities:
+                self.delete(entity)
+
+        def delete_all(self):
+            self._engine.execute(entity_table.delete())
+            self._identity_map = {}
+            for session in self._sessions.values():
+                session.close()
+            self._sessions = {}
+
+        def delete_by_id(self, id_: K):
+            if not isinstance(id_, tuple):
```

Comment:
```suggestion
            if not isinstance(id_, collections.Iterable):
```
кажется будет полезна возможность передать и list и set

Reply:
Set для композитного первичного ключа - это вроде чересчур :) даже (1,2) или (2,1) есть же разница. Тем более (1,1) в множестве вообще схлопнется.

Reply:
я не правильно код понял изначально